### PR TITLE
Updates and fixes for ramp fitting for recent PRs and Issues.

### LINKED ISSUE #2848: Ramp-fitting returning zero for all background pixels
Issue [JP-453](https://jira.stsci.edu/browse/JP-453) was created by Bryan Hilbert:

I have created some simulated NIRCam data, and am getting results out of the ramp-fitting step that I don't understand. I'm getting slopes of exactly zero for all of my background pixels.

The simulated data that I'm looking at is an exposure composed of two integrations of only 3 groups each. When I look at the output from the jump step, things look reasonable. The number of flagged jumps across the detector appears relatively normal. The example pixel I've followed through the pipeline steps has signal values of [14.4, 9.98. 4.89] in the three groups of the first integration. There are no flagged jumps. But in the rate image from the ramp-fitting step, it has a value of 0. The ramp also contains a number of very bright stars, leading to some pixels which are saturated in all groups (including the zeroframe). Those pixels also have a slope of 0, as expected. I haven't yet investigated to see if the high-signal pixels have reasonable slope values.

I updated my environment to use today's dev version of the pipeline for this testing.  0.13.0a.dev146+g4abefa32

I have attached the very simple script I used to call the pipeline while investigating this. The file I'm testing with is located in: /grp/jwst/wit/nircam/hilbert/simulated_ramps/car-nrc-19/simulated_data/jw01068004001_0110s_00001_nrcb5_super_jump.fits (the uncal version, as well as the outputs from several other pipeline steps are present in there as well).

### PR DESCRIPTION
**Pull request by dmggh**
Fixes #2848, #2233, #2885, #2832

### CODE REVIEW

**Comment by dmggh** [comment]
Details of fixes and updates:
1) "Ramp-fitting returning zero for all background pixels"; Issue #2848, JP-453.

2) "MIRI ramps with jumps flagged at group 2 result in slopes of 0 in the rate image"; Issue #2233,

3) "Processing pixels in ramp fitting in which all groups are saturated"; Issue #2885.

4) "Ramp Fit fails when only two groups are in a segment after cosmic ray hits."; Issue #2832, JP-450.

5) Fixed a bug in which the keywords from the input were not included in the OPT output header.

6) Simplified and clarified classification of segment types based on DQ flags.

7) Added handling of ramps ending in 2 saturated groups.

8) "Fix units for Read Noise Variance in ramp_fit (PR #2767)". This may needed to revised based on Mike Regan's comment when he closed this PR.

9) Added check to handle integration-specific variances for too short segments.

10) Removed some obsolete code

File: jwst/ramp_fitting/ramp_fit.py [hunk 1]
```diff
@@ -174,6 +174,18 @@ def ols_ramp_fit(model, buffsize, save_opt, readnoise_model, gain_model,
...
+            #   remove the JUMP_DET flag from the group dq for those pixels so
+            #   that those groups will be included in the fit.
+            wh_cr = np.where( np.bitwise_and(model.groupdq[:,0,:,:],
+                              dqflags.group['JUMP_DET']) != 0 ) 
+            num_cr_1st = len(wh_cr[0])
+
+            for ii in range(num_cr_1st):
+                model.groupdq[ wh_cr[0][ii], 0, wh_cr[1][ii],
+                    wh_cr[2][ii]] -=  dqflags.group['JUMP_DET']
+                print (' ii = ', ii)
```
Comment (philhodge) [inline jwst/ramp_fitting/ramp_fit.py]: This print statement should be deleted.

File: jwst/ramp_fitting/ramp_fit.py [hunk 2]
```diff
@@ -202,12 +214,24 @@ def ols_ramp_fit(model, buffsize, save_opt, readnoise_model, gain_model,
...
     #   frames averaged on-board for a group, i.e., it does not include the
     #   groupgap.
-    effintim, nframes, groupgap, dropframes1= utils.get_efftim_ped(model)
+    effintim, nframes, groupgap, dropframes1 = utils.get_efftim_ped(model)
 
     # Get GROUP DQ and ERR arrays from input file
     gdq_cube = model.groupdq
     gdq_cube_shape = gdq_cube.shape
 
+    # If the all pixels have their initial groups flagged as saturated, the DQ
```
Comment (philhodge) [inline jwst/ramp_fitting/ramp_fit.py]: "If the all pixels ..." --> "If all the pixels ..."

File: jwst/ramp_fitting/ramp_fit.py [hunk 3]
```diff
@@ -219,24 +243,22 @@ def ols_ramp_fit(model, buffsize, save_opt, readnoise_model, gain_model,
...
     nrows = calc_nrows(model, buffsize, cubeshape, nreads)
 
     # Get readnoise array for calculation of variance of noiseless ramps, and
     #   gain array in case optimal weighting is to be done
     readnoise_2d, gain_2d = utils.get_ref_subs(model, readnoise_model,
                                                gain_model, nframes)
 
-    # Flag any bad pixels in the gain
-    pixeldq = utils.reset_bad_gain( pixeldq, gain_2d )
+    # Get Pixel DQ array from input file. The incoming RampModel has uint8
```
Comment (philhodge) [inline jwst/ramp_fitting/ramp_fit.py]: The pixeldq in RampModel is uint32, so this should be OK to start with.  (It's the groupdq that's uint8 in RampModel.)

File: jwst/ramp_fitting/ramp_fit.py [hunk 4]
```diff
@@ -1502,12 +1551,17 @@ def fit_next_segment(start, end_st, end_heads, pixel_done, data_sect, mask_2d,
     #    - set number of end to 0
     #    - add slopes and variances to running sums
     #    - set pixel_done to True to designate all fitting done
+    #  For segments of this type, the final good group in the segment is a CR
+    #    and/or SAT and is not the final group in the ramp, and the variable
```
Comment (philhodge) [inline jwst/ramp_fitting/ramp_fit.py]: Can the final group in a segment be a CR?  Why wouldn't that start a new segment?  Also, how can a group be good if it's either affected by a CR or is saturated?

File: jwst/ramp_fitting/ramp_fit.py [hunk 5]
```diff
@@ -2237,14 +2283,7 @@ def calc_num_seg(gdq, n_int):
...
 
-    # Do not want to return a value > the number of groups, which can occur if 
-    #  this is a MIRI dataset in which the first or last group was flagged as 
-    #  DO_NOT_USE and also flagged as a jump. 
-    max_num_seg = int(max_cr) +1 
-    if (max_num_seg > gdq.shape[1]): 
-        max_num_seg = gdq.shape[1]
-
-    return max_num_seg
+    return int(max_cr) +1 # n CRS implies n+1 segments
```
Comment (philhodge) [inline jwst/ramp_fitting/ramp_fit.py]: This looks like the previous version of this function.

File: jwst/ramp_fitting/utils.py [hunk 6]
```diff
@@ -278,6 +278,9 @@ def output_optional(self, model, effintim):
         self.var_p_seg[self.var_p_seg > 0.4 * LARGE_VARIANCE ] = 0.
         self.var_r_seg[self.var_r_seg > 0.4 * LARGE_VARIANCE ] = 0.
 
+        # Tiny 'weights' values correspond to non-existent segments, so set to 0.
+        self.weights[1./self.weights > LARGE_VARIANCE ] = 0.
```
Comment (philhodge) [inline jwst/ramp_fitting/utils.py]: In other cases, you compare with `0.4 * LARGE_VARIANCE`, but here you compare with just `LARGE_VARIANCE`.  Is that OK?  See line 462 in ramp_fit.py, for example.

**Pull request merged**